### PR TITLE
Fix test + remove buffer options + return nb pages

### PR DESCRIPTION
**Pull request by Atinux**
Hello!

Thanks for your awesome node module!
I removed the buffer option in this version (already deprecated in
0.3.0) + the script return a JSON with the number of pages in the PDF!

I also updated the README and the documentation!

I you like it, don’t forget to `npm publish` :)

Regards,
Sebastien

### CODE REVIEW

**Comment by marcbachmann** [comment]
Sorry, I haven't set up appveyor or travis correctly. I'll have a detailed look into your pull-request somewhen this week.

**Comment by Atinux** [comment]
Thank you :)

**Comment by Atinux** [comment]
I added the tests for Travis and AppVeyor too!

File: Changelog.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+0.4.0
```
Comment (marcbachmann) [inline Changelog.md]: I suggest you to do version bumps always in the master branch after a feature is merged. It looks strange if there are that many other commits after creating a changelog with a specific version.

File: README.md [hunk 2]
```diff
@@ -11,9 +11,15 @@
 var fs = require('fs');
 var pdf = require('./lib');
 var html = fs.readFileSync('./test/businesscard.html', 'utf8')
-pdf.create(html, { width: '50mm', height: '90mm'}, function(err, buffer) {
+pdf.create(html, { filename: './businesscard.pdf', width: '50mm', height: '90mm'}, function(err, res) {
```
Comment (marcbachmann) [inline README.md]: :+1: for the filename in the example. We should also use a realistic size here.

File: README.md [hunk 3]
```diff
@@ -11,9 +11,15 @@
...
   if (err) return console.log(err);
-  fs.writeFile('businesscard.pdf', buffer);
+  console.log(res);
+  /*
+    {
+      filename: './businesscard.pdf',
```
Comment (marcbachmann) [inline README.md]: Oh, we should improve that. The response should be an absolute path. We should parse an input filename to be absolute. e.g. call `options.filename = path.resolve(options.filename)` in the `create` method.

File: lib/index.coffee [hunk 4]
```diff
@@ -49,21 +49,11 @@ exports.create = (string, options, callback) ->
...
-        return callback(err) if err
-        fs.unlink filename, (err) ->
-          callback(err, buffer)
-
-    else
-      callback(null, file)
 
+    try
+      file = JSON.parse(file)
+    catch err
```
Comment (marcbachmann) [inline lib/index.coffee]: That's a wrong behaviour. We should return the error in the callback if there's one.

File: lib/scripts/pdf_a4_portrait.coffee [hunk 5]
```diff
@@ -21,6 +21,7 @@ exit('Did not receive any html') if !json.html?.trim()
 options = json.options
 page = webpage.create()
 page.content = json.html
+nbPages = 0
```
Comment (marcbachmann) [inline lib/scripts/pdf_a4_portrait.coffee]: What does `nb` stand for?

**Comment by marcbachmann** [comment]
I'd like to have the pages in the response. But sadly this only works if we define a header or footer because we only execute that method when rendering headers or footers. That might look buggy.
What do you think about that?

The travis config is also ok to merge. Can you create a new pull-request only with the trevis file? I can merge that directly. I'll implement appveyor separately (for windows tests). 

With the other things i'd like to wait because I've changed the api a bit.

File: Changelog.md (hunk 1 above)
Comment (Atinux) [inline Changelog.md]: You right!

File: README.md (hunk 3 above)
Comment (Atinux) [inline README.md]: Added in https://github.com/Atinux/node-html-pdf/commit/e24ebfa3aa5aaa7b94920e4ff2d1de351ab3e1c8

File: lib/index.coffee (hunk 4 above)
Comment (Atinux) [inline lib/index.coffee]: That's what I thought first.
But if you use a custom script that return only text but not JSON?
This functionality is just a shortcut if the result send back from PhantomJS can be parsed into an object or not.
For example, I'm using the default script, but also another custom script which send back a number, I don't want to have an error if JSON.parse(5) fail for example.

File: lib/scripts/pdf_a4_portrait.coffee (hunk 5 above)
Comment (Atinux) [inline lib/scripts/pdf_a4_portrait.coffee]: I changed it.
The only problem with this, is if you did not have an header or footer, you will not have the right number of pages :/

**Comment by Atinux** [comment]
I agree, this is annoying that we cannot have the right number of pages without a header or footer given.

I think you can forget this pull request if you changed the API, just maybe be inspired for futur changes by what I've done.

:)

**Pull request closed without merging**